src/graphql/graphql.ts

```diff
@@ -1667,3 +1696,62 @@ function buildRefreshedFilter(
     },
   };
 }
+
+class PrimaryKeyResolver {
+  constructor(
+    readonly graphQLSchema: gql.GraphQLSchema,
+    readonly primaryKeys: Dictionary<ReadonlyArray<string>>,
+    readonly isTopLevelModelTypeChecker = isV1ModelType
+  ) {}
+
+  /**
+   * Fully resolves the primary keys of models in the schema.
+   *
+   * E.g., if the primary keys of 'cicd_Organization' are 'source'
+   * and 'uid' (both scalars) and the primary keys of 'cicd_Pipeline'
+   * are 'organization' (foreign key to cicd_Organization) and 'uid' (scalar),
+   * the fully resolved primary key of 'cicd_Pipeline' is
+   * { organization { source uid } uid }
+  */
+  public resolvePrimaryKeys(): Dictionary<string> {
+      const result: Dictionary<string> = {};
+
+      for (const name of Object.keys(this.graphQLSchema.getTypeMap())) {
+        const type = this.graphQLSchema.getType(name);
+        if (this.isTopLevelModelTypeChecker(type)) {
+          result[name] = this.resolvePrimaryKey(type);
+        }
+      }
+
+      return result;
+  }
+
+  @Memoize()
```

Comment:
Add type to memorize key

Reply:
From the [docs](https://www.npmjs.com/package/typescript-memoize) I believe it's not needed for single param methods?

Reply:
Since your parameter is not a simple one, I would add a custom hash function

Reply:
Done, thanks!